hydrus/data/resource_based_classes.py

```diff
@@ -0,0 +1,209 @@
+    for fk in foreign_keys:
+        # the name of the column through which this foreign key relationship
+        # is being established
+        fk_column = fk.info["column_name"]
+        try:
+            fk_object = object_[fk_column]
+        except KeyError as e:
+            wrong_property = e.args[0]
+            raise PropertyNotGiven(type_=wrong_property)
+        # insert the foreign key object
+        fk_object_id = insert_object(fk_object, session)
+        # put the id of the foreign instance in this table's column
+        object_[fk_column] = fk_object_id
+    try:
+        # remove the @type from object before using the object to make a
+        # instance of it using sqlalchemy class
+        object_.pop("@type")
+        inserted_object = database_class(**object_)
+    except TypeError as e:
+        # extract the wrong property name from TypeError object
+        wrong_property = e.args[0].split("'")[1]
+        raise PropertyNotFound(type_=wrong_property)
+    try:
+        session.add(inserted_object)
+        session.commit()
+    except Exception as e:
+        # catching any database contraint errors
+        session.rollback()
+        contraint_error = e.orig
+        raise DatabaseConstraintError(contraint_error)
+    return inserted_object.id
+
+
+def get_object(
+    query_info: Dict[str, str], session: scoped_session
+) -> Dict[str, str]:
+    """
+    Get the object from the database
+    :param query_info: Dict containing the id and @type of object that has to retrieved
+    :param session: sqlalchemy session
+    :return: dict of object with its properties
+    """
+    type_ = query_info["@type"]
+    id_ = query_info["id_"]
+    database_class = get_database_class(type_)
+    try:
+        object_ = (
+            session.query(database_class)
+            .filter(database_class.id == id_)
+            .one()
+        ).__dict__
+    except NoResultFound:
+        raise InstanceNotFound(type_=type_, id_=id_)
+    # Remove the unnecessary keys from the object retrieved from database
+    object_template = copy.deepcopy(object_)
+    object_template.pop("_sa_instance_state")
+    object_template.pop("id")
+    object_template["@type"] = query_info["@type"]
+    return object_template
+
+
+def delete_object(query_info: Dict[str, str], session: scoped_session) -> None:
+    """
+    Delete the object from the database
+    :param query_info: Dict containing the id and @type of object that has to retrieved
+    :param session: sqlalchemy session
+    """
+    type_ = query_info["@type"]
+    id_ = query_info["id_"]
+    database_class = get_database_class(type_)
+    try:
+        object_ = (
+            session.query(database_class)
+            .filter(database_class.id == id_)
+            .one()
+        )
+    except NoResultFound:
+        raise InstanceNotFound(type_=type_, id_=id_)
+    session.delete(object_)
+    try:
+        session.commit()
+    except Exception:
```

Comment:
It's not a good idea to use generic exceptions when using an external library. Try and figure out what might go wrong in SQLAlchemy and catch those specific exceptions.

Reply:
Yes, I understand. I will not do that next time :sweat_smile: